Add frontend implementation for Reviewed! (#2)
ballerina-guides/wso2con-24: session-3/frontend/README.md

# Frontend implementation for Reviewed!

## Overview

This is the frontend implementation for Reviewed!. It consists of a React application that works with a Ballerina GraphQL backend. It uses the Apollo Client on the frontend to connect to the GraphQL backend.

## Get Started

There are two ways to instantly serve up the frontend application:

1. Serve the application using NodeJS.

```bash
npx serve dist --single --listen 3000
```

2. Serve the application using Python.

```bash
python -m http.server 3000 --directory dist
```

> > Note: URL rewriting is not supported when serving the production build using python. (contents of ErrorPage.jsx will not render for invalid URLs)

Reviewer: ```suggestion
> Note: URL rewriting is not supported when serving the production build using Python. Also, the contents of ErrorPage.jsx will not render for invalid URLs.
```

- Open the browser and navigate to: [http://localhost:3000](http://localhost:3000)

## Running the Application in Development Mode

### Prerequisites

Install the following software.

> - [Node.js version: v20.12.0](https://nodejs.org/en/blog/release/v20.12.0)

Reviewer: ```suggestion
- [Node.js (version v20.12.0)](https://nodejs.org/en/blog/release/v20.12.0)
```

For consistency.

- [npm (version 10.5.0 or later)](https://www.npmjs.com/package/npm)

### Steps

1. Install the required project dependencies by running the following command.

```bash
npm install
```

> 2. For changing the backend endpoints, update the environment variables in the `.env` file.

Reviewer: ```suggestion
2. To change the backend endpoints, update the environment variables in the `.env` file.
```

> ```bash

Reviewer: ````suggestion
```
````

VITE_GRAPHQL_ENDPOINT_HTTP=http://localhost:9000/reviewed
VITE_GRAPHQL_ENDPOINT_WS=ws://localhost:9000/reviewed
```

3. Start the application.

```bash'
npm run dev
```

> 4. Open the browser and navigate to [http://localhost:3000](http://localhost:3000).
> 
> 5. For deploying the application, create a production build by running the following command.

Reviewer: ```suggestion
4. Open the browser and navigate to [http://localhost:3000](http://localhost:3000).

5. To deploy the application, create a production build by running the following command.
```

> 5. For deploying the application, create a production build by running the following command.

Author: Since this section is in the context of a user who is running the front-end in dev mode, deployment was mentioned as the last point. The reason is that during development, the changes made in the JavaScript files of the React code will be instantly visible on the served URL. However if the user has created a production build and is serving it instead, the changes made on the JavaScript files won't be visible on the served files.

# Frontend implementation for Reviewed!

## Overview

This is the frontend implementation for Reviewed!. It consists of a React application that consumes Ballerina GraphQL services using Apollo Client.

## Prerequisites

- [Node.js version: v20.12.0](https://nodejs.org/en/blog/release/v20.12.0)

> - npm (version 10.5.0 or later)

Reviewer: Add a link for consistency.

> - npm (version 10.5.0 or later)

Reviewer: With the two approaches, we don't need this for both do we? 

I think we should first say there are two ways to spin this up, introduce them, and later organize sections accordingly.

## Getting Started

1. Install the required dependencies by running the following command:

```bash
npm install
```

> 2. Create a `.env` file and add the following environment variables:

Reviewer: I think we can add this by default and ask to modify if required.

```bash
VITE_GRAPHQL_ENDPOINT_HTTP=https://localhost:9000/reviewed
VITE_GRAPHQL_ENDPOINT_WS=wss://localhost:9000/reviewed
```

3. Start the frontend application by running the following command:

```bash'
npm run dev
```

4. Open the browser and navigate to: [http://localhost:3000](http://localhost:3000)

## Deployment

1. Create a production build by running the following command (Optional):

```bash
npm run build
```

2. Serve the production build by running the following command:

- Serving the production build using NodeJS:

  - Run the following command:

  ```bash
  npm run preview
  ```

  - Open the browser and navigate to: [http://localhost:3000](http://localhost:3000)

- Serving the production build using python:

  - Run the following command:

  ```bash
  py -m http.server 3000 --directory dist
  ```

  - Open the browser and navigate to: [http://localhost:3000](http://localhost:3000)

>   > Note: URL rewriting is not supported when serving the production build using python. (contents of ErrorPage.jsx will not render for invalid URLs)

Reviewer: ```suggestion
  > Note: URL rewriting is not supported when serving the production build using Python. (contents of ErrorPage.jsx will not render for invalid URLs)
```

Isn't the latter part a separate point?
